snapshots/windows/cimfs.go

```diff
@@ -0,0 +1,276 @@
+	}
+	return s.createCimfsMounts(ctx, m, key, opts...)
+}
+
+func (s *cimfsSnapshotter) createCimfsMounts(ctx context.Context, m []mount.Mount, key string, opts ...snapshots.Opt) (_ []mount.Mount, retErr error) {
+	if len(m) != 1 {
+		return m, fmt.Errorf("expected exactly 1 mount from legacy windows snapshotter, found %d", len(m))
+	}
+	m[0].Type = "cimfs"
+	return m, nil
+}
+
+// Mounts returns the cimfs mounts for the snapshot identified by key.
+//
+// This can be used to recover mounts after calling View or Prepare.
+func (s *cimfsSnapshotter) Mounts(ctx context.Context, key string) ([]mount.Mount, error) {
+	ctx, t, err := s.legacySn.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return nil, err
+	}
+	defer t.Rollback()
+
+	sn, err := storage.GetSnapshot(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("get snapshot: %w", err)
+	}
+
+	mounts, err := s.legacySn.mounts(sn, key), nil
+	if err != nil {
+		return nil, err
+	}
+	return s.createCimfsMounts(ctx, mounts, key)
+}
+
+func (s *cimfsSnapshotter) Commit(ctx context.Context, name, key string, opts ...snapshots.Opt) error {
+	ctx, t, err := s.legacySn.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil {
+			if rerr := t.Rollback(); rerr != nil {
+				log.G(ctx).WithError(rerr).Warn("failed to rollback transaction")
+			}
+		}
+	}()
+
+	// grab the existing id
+	id, _, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	usage, err := fs.DiskUsage(ctx, s.legacySn.getSnapshotDir(id))
+	if err != nil {
+		return err
+	}
+	if ok, err := s.isCimLayer(ctx, key); err != nil {
+		return err
+	} else if ok {
+		cimUsage, err := cimfs.GetCimUsage(ctx, getCimLayerPath(s.cimDir, id))
+		if err != nil {
+			return err
+		}
+		usage.Size += int64(cimUsage)
+	}
+
+	if _, err = storage.CommitActive(ctx, key, name, snapshots.Usage(usage), opts...); err != nil {
+		return fmt.Errorf("commit snapshot: %w", err)
+	}
+	return t.Commit()
+}
+
+// Remove abandons the transaction identified by key. All resources
+// associated with the key will be removed.
+func (s *cimfsSnapshotter) Remove(ctx context.Context, key string) error {
+	ctx, t, err := s.legacySn.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return err
+	}
+	defer t.Rollback()
+
+	// Get info before calling legacySn.Remove
+	id, _, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return fmt.Errorf("%w: get snapshot info %s", errdefs.ErrFailedPrecondition, err)
+	}
+
+	if isCimLayer, err := s.isCimLayer(ctx, key); err != nil {
+		return fmt.Errorf("%w: detect if cim layer: %s", errdefs.ErrFailedPrecondition, err)
+	} else if isCimLayer {
+		if err := cimlayer.DestroyCimLayer(s.legacySn.getSnapshotDir(id)); err != nil {
+			return fmt.Errorf("%w: destroy cim layer: %s", errdefs.ErrFailedPrecondition, err)
+		}
+	}
+	// Must rollback this transaction before calling legacySn.Remove since that function opens its
+	// own transaction.
+	t.Rollback()
```

Comment:
I haven't looked at the behavior of Rollback here. What happens if this fails and you call Remove?

Reply:
This code is removed after refactoring.